### PR TITLE
Docker: support passing additional commandline args to synapse

### LINKED ISSUE #6662: Support config fragments for docker image
**Description:**

I'm using the very nice docker image of synapse and would like to follow the default configuration and only override specific parts of it. It seems that synapse supports config fragments in a directory (which the debian package seems to be [using](https://github.com/matrix-org/synapse/blob/master/debian/matrix-synapse.service#L10) ), but the docker image runs [start.py](https://github.com/matrix-org/synapse/blob/master/docker/start.py#L250) which doesn't have the relevant flags and doesn't allow passing them in.

### PR DESCRIPTION
**Pull request by samuel-p**
Updated the docker `start.py` to allow passing commandline args to the synapse process (fixes #6662).

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)
* [x] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

`Signed-off-by: Samuel Philipp <codes@samuel-philipp.de>`

### CODE REVIEW

File: docker/start.py [hunk 1]
```diff
@@ -205,36 +205,43 @@ def main(args, environ):
             config_dir, config_path, environ, ownership
         )
 
-    if mode is not None:
-        error("Unknown execution mode '%s'" % (mode,))
+    args = ["python"] + args[1:]
```
Comment (richvdh) [inline docker/start.py:208]: I don't think it is a good idea to conflate the "mode" and the synapse parameters in this way.  I suggest adding an explicit `mode` such as `run` which must be specified before any synapse parameters.

File: docker/start.py [hunk 2]
```diff
@@ -205,36 +205,43 @@ def main(args, environ):
...
 
-    config_dir = environ.get("SYNAPSE_CONFIG_DIR", "/data")
-    config_path = environ.get("SYNAPSE_CONFIG_PATH", config_dir + "/homeserver.yaml")
+    if "-m" not in args:
+        args = ["-m", synapse_worker] + args
```
Comment (richvdh) [inline docker/start.py:211]: you're putting ` ["-m", synapse_worker]` *before* `python`?

File: docker/start.py [hunk 3]
```diff
@@ -205,36 +205,43 @@ def main(args, environ):
...
             )
 
-        error(
-            "Config file '%s' does not exist. You should either create a new "
-            "config file by running with the `generate` argument (and then edit "
-            "the resulting file before restarting) or specify the path to an "
-            "existing config file with the SYNAPSE_CONFIG_PATH variable."
-            % (config_path,)
-        )
+        if "--config-path=" + config_path not in args:
```
Comment (richvdh) [inline docker/start.py:240]: why do we need this condition? doesn't the condition at line 213 make it redundant?

File: docker/start.py [hunk 4]
```diff
@@ -205,36 +205,43 @@ def main(args, environ):
...
-    config_path = environ.get("SYNAPSE_CONFIG_PATH", config_dir + "/homeserver.yaml")
+    if "-m" not in args:
+        args = ["-m", synapse_worker] + args
 
-    if not os.path.exists(config_path):
-        if "SYNAPSE_SERVER_NAME" in environ:
-            error(
-                """\
-Config file '%s' does not exist.
+    if not filter(lambda p: p.startswith("--config-path"), args):
```
Comment (richvdh) [inline docker/start.py:213]: some comments would be nice here. what are we doing, and why?

Comment (richvdh) [inline docker/start.py:213]: I think `any` is clearer here:

```suggestion
    if not any(p.startswith("--config-path") for p in args):
```

File: changelog.d/8390.docker [hunk 5]
```diff
@@ -0,0 +1 @@
+Added support for passing commandline args to the synapse process. Contributed by @samuel-p.
```
Comment (richvdh) [inline changelog.d/8390.docker:1]:
```suggestion
Add support for passing commandline args to the synapse process. Contributed by @samuel-p.
```

**Review by richvdh — CHANGES_REQUESTED** [review]
I think this looks like a great start - thanks very much!

I've made a few comments below. Please also make sure that you update the documentation to match - in particular https://github.com/matrix-org/synapse/blob/develop/docker/README.md.

File: docker/start.py (hunk 1 above)
Comment (samuel-p) [reply docker/start.py:208]: Yes, also thought of this. Can change that.

File: docker/start.py (hunk 2 above)
Comment (samuel-p) [reply docker/start.py]: indeed, fixed that

File: docker/start.py (hunk 3 above)
Comment (samuel-p) [reply docker/start.py]: yes, removed it

File: docker/start.py (hunk 4 above)
Comment (samuel-p) [reply docker/start.py]: added it

**Comment by samuel-p** [comment]
Thanks for your comments. I hope I could resolve them. I also updated the docker docs. Please check again, if something else is missing.

File: docker/start.py [hunk 6]
```diff
@@ -205,41 +205,53 @@ def main(args, environ):
             config_dir, config_path, environ, ownership
         )
 
-    if mode is not None:
-        error("Unknown execution mode '%s'" % (mode,))
+    if mode == "run":
```
Comment (richvdh) [inline docker/start.py:208]: I suggest you keep this as an error check, to save having a massive `if` statement:

```suggestion
    if mode != "run":
        error("Unknown execution mode '%s'" % (mode,))
```

File: docker/start.py [hunk 7]
```diff
@@ -205,41 +205,53 @@ def main(args, environ):
...
+        if "-m" not in args:
+            args = ["-m", synapse_worker] + args
 
-    if not os.path.exists(config_path):
-        if "SYNAPSE_SERVER_NAME" in environ:
-            error(
-                """\
-Config file '%s' does not exist.
+        # if there are no config files passed to synapse, try adding the default file
+        if not any(p.startswith("--config-path") for p in args):
```
Comment (richvdh) [inline docker/start.py:215]: we should probably also check for `-c` here.

File: docker/README.md [hunk 8]
```diff
@@ -94,6 +94,18 @@ The following environment variables are supported in run mode:
...
+```
+docker run -d --name synapse \
+    --mount type=volume,src=synapse-data,dst=/data \
+    -p 8008:8008 \
+    matrixdotorg/synapse:latest run \
+    -m synapse.app.generic_worker \
+    --config-path=/data/homeserver.yaml \
+    --config-path=/data/generic_worker.yaml
+```
+
```
Comment (richvdh) [inline docker/README.md:108]: can you document that `--config-path` is set to `SYNAPSE_CONFIG_PATH` unless it is set explicitly?

**Review by richvdh — CHANGES_REQUESTED** [review]
thanks for this, and sorry for taking a while to get back to it.

I've got a couple more comments below.

File: docker/start.py (hunk 7 above)
Comment (samuel-p) [reply docker/start.py]: Yes, missed that.

File: docker/README.md (hunk 8 above)
Comment (samuel-p) [reply docker/README.md:108]: Yes, also mentioned that `SYNAPSE_WORKER` is used, if `-m` is not set.

**Review by richvdh — APPROVED** [review]
lgtm!

**Comment by richvdh** [comment]
Thank you for your work on this!

**Pull request merged**